### PR TITLE
Test framework

### PR DESCRIPTION
**Pull request by mike-sul**

### CODE REVIEW

**Comment by mike-sul** [comment]
@doanac @detsch This PR includes all I planned to add to the "test framework" during this initial implementation iteration.

File: test/integration/smoke_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	runCmd := func(t *testing.T, args ...string) []byte {
+		c := exec.Command(composeExec, args...)
+		c.Dir = appDir
+		output, err := c.CombinedOutput()
+		if err != nil {
+			t.Errorf("failed to run `%s` command: %s\n", args[0], output)
```
Comment (doanac) [inline test/integration/smoke_test.go]: You probably want `t.Fatal` here or your test is just going to fail everything after this. It's also going to get tedious checking for errors everywhere. The `require.Nil` package/function makes this a lot easier

Comment (doanac) [reply test/integration/smoke_test.go]: I get the feeling all your tests are going to need this function. I'm guessing the code is going to be easier if you make this a function

File: test/fixtures/images/runner/forever.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import (
+	"time"
+)
+
+func main() {
+	for {
+		time.Sleep(1 * time.Minute) // Sleep for 1 hour in each iteration
+	}
+}
```
Comment (doanac) [inline test/fixtures/images/runner/forever.go:11]: why not just use an alpine container who's entrypoint is `sleep 1000d`?

File: test/fixtures/composectl_cmds.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
+package fixtures
```
Comment (doanac) [inline test/fixtures/composectl_cmds.go:1]: I think you'll want this file name to end with `_test.go` to make sure people don't try and use this as part of you API.

**Review by doanac — COMMENTED** [review]
I think it would be easier to introduce the test fixture utilities at the first and then do the test files. It would make the PR a little easier to focus on what you are testing

**Comment by mike-sul** [comment]
> I think it would be easier to introduce the test fixture utilities at the first and then do the test files.

I suppose it makes sense from the test case perspective, and a commit history will be cleaner. I will update PR accordingly. (I just wanted to justify the fixtures and how they evolutionary appears)

File: test/fixtures/composectl_cmds.go (hunk 3 above)
Comment (mike-sul) [reply test/fixtures/composectl_cmds.go:1]: I assume the package name, i.e. `fixtures` indicates that this is not API.

File: test/fixtures/images/runner/forever.go (hunk 2 above)
Comment (mike-sul) [reply test/fixtures/images/runner/forever.go:11]: I'd like to test pulling images from our Registry. And the most of all, it is really easy to hit the docker.io daily limit if/when the tests are executed quite often, which is the case during development phase.

File: test/integration/smoke_test.go (hunk 1 above)
Comment (mike-sul) [reply test/integration/smoke_test.go]: Let me restructure the PR so the fixtures appear at first and then the test cases. Probably, after that the comments will become irrelevant.

Comment (mike-sul) [reply test/integration/smoke_test.go]: >The require.Nil package/function makes this a lot easier

Agree, the simple function should do the trick, not sure it is worth pulling another deps for it though.

File: test/fixtures/images/runner/forever.go (hunk 2 above)
Comment (mike-sul) [reply test/fixtures/images/runner/forever.go:11]: Probably, we could pull the alpine image or even something smaller from docker.io once push it to the registry running withing the compose test env.

File: test/fixtures/composectl_cmds.go (hunk 3 above)
Comment (mike-sul) [reply test/fixtures/composectl_cmds.go:1]: In the end, a client make use of these fixtures in their tests.

Comment (doanac) [reply test/fixtures/composectl_cmds.go:1]: i think a client could for their tests if it ends in _test.go. But making _test.go makes this explicit and makes sure it doesn't contribute to the final binary size.

Comment (mike-sul) [reply test/fixtures/composectl_cmds.go:1]: It doesn't contribute to the final binary size.

It makes sense to add suffix `_test.go` to the files containing test fixtures and test code if they are located in the same directory/package as the files of the code under the test. In this case compiler excludes `*_test.go` files from binary even if they are part of the code package.

In this particular case, fixtures are placed into the dedicated directory/package and are not compiled unless explicitly imported.

File: test/integration/smoke_test.go (hunk 1 above)
Comment (mike-sul) [reply test/integration/smoke_test.go]: @doanac I've restructured commits in the PR. Please, review it now.

File: test/compose/registry/certs/registry.crt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply test/compose/registry/certs/registry.crt:1]: Added the script https://github.com/foundriesio/composeapp/pull/28/commits/b9366d31df97bb9cdb39086760cad455a0674e74#diff-f54df8e08c3b21fa819dac5c97750f3835e5ae32da8e4e02d18e6eab35a01608

**Review by doanac — APPROVED** [review]

**Comment by detsch** [comment]
Perhaps adding a comment to the README.md mentioning how to execute the e2e tests locally would be helpful, which seems to be:
```
./dev-shell.sh
make test-e2e
```
Other then that, LGTM.

**Review by detsch — APPROVED** [review]

**Comment by mike-sul** [comment]
> Perhaps adding a comment to the README.md mentioning how to execute the e2e tests locally would be helpful, which seems to be:
> 
> ```
> ./dev-shell.sh
> make test-e2e
> ```

It makes sense, I'll update the readme.

**Comment by mike-sul** [comment]
@detsch https://github.com/foundriesio/composeapp/pull/28/commits/6eca2756071f7d600ad1f2f4ff0227147a9b78c2

**Comment by detsch** [comment]
> @detsch [6eca275](https://github.com/foundriesio/composeapp/commit/6eca2756071f7d600ad1f2f4ff0227147a9b78c2)

Looks good

**Pull request merged**